writer.go

```diff
@@ -183,3 +180,73 @@ func trimPineconeKey(key string) (trimmed string, hasPrefix bool) {
 
 	return key, false
 }
+
+type recordBatch interface {
+	writeBatch(context.Context, *Writer) (int, error)
+}
+
+type upsertBatch struct {
+	vectors []*pinecone.Vector
+}
+
+func (b upsertBatch) writeBatch(ctx context.Context, writer *Writer) (int, error) {
+	return writer.UpsertVectors(ctx, b.vectors)
+}
+
+type deleteBatch struct {
+	ids []string
+}
+
+func (b deleteBatch) writeBatch(ctx context.Context, writer *Writer) (int, error) {
+	err := writer.DeleteVectorsByID(ctx, b.ids)
+	if err != nil {
+		return 0, err
+	}
+
+	return len(b.ids), nil
+}
+
+// parseRecords processes a slice of records and groups them into batches based
+// on their operation type. New batches are started whenever the operation type
+// switches from upsert to delete or vice versa.
+// Records are batched this way so that we preserve conduit's requirement of writing
+// records sequentially.
+func parseRecords(records []sdk.Record) ([]recordBatch, error) {
```

Comment:
Alternatively, we can also call it `make/write/buildBatches`, since the focus I believe should be on batches.